drivers/gpio/gpio_pcal6416a.c

```diff
@@ -0,0 +1,269 @@
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_OUTPUT_PORT_1,
+				 (uint8_t)(pins_cfg->outputs_high >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_CONFIGURATION_0,
+				 (uint8_t)pins_cfg->configured_as_inputs);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_CONFIGURATION_1,
+				 (uint8_t)(pins_cfg->configured_as_inputs >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int pcal6416a_inputs_read(const struct i2c_dt_spec *i2c, uint16_t *int_sources,
+				 uint16_t *input_port)
+{
+	int rc;
+	uint8_t value_low;
+	uint8_t value_high;
+
+	rc = pcal6416a_i2c_read(i2c, PCAL6416A_REG_INTERRUPT_STATUS_0, &value_low);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_read(i2c, PCAL6416A_REG_INTERRUPT_STATUS_1, &value_high);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	*int_sources = value_low | (value_high << 8);
+
+	/* This read also clears the generated interrupt if any. */
+	rc = pcal6416a_i2c_read(i2c, PCAL6416A_REG_INPUT_PORT_0, &value_low);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_read(i2c, PCAL6416A_REG_INPUT_PORT_1, &value_high);
+	if (rc != 0) {
+		LOG_ERR("failed to read input port: %d", rc);
+		return -EIO;
+	}
+
+	*input_port = value_low | (value_high << 8);
+
+	return 0;
+}
+
+static int pcal6416a_outputs_write(const struct i2c_dt_spec *i2c, uint16_t outputs)
+{
+	int rc;
+
+	/*
+	 * No need to limit `out` to only pins configured as outputs,
+	 * as the chip anyway ignores all other bits in the register.
+	 */
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_OUTPUT_PORT_0, (uint8_t)outputs);
+
+	if (rc != 0) {
+		LOG_ERR("failed to write output port: %d", rc);
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_OUTPUT_PORT_1, (uint8_t)(outputs >> 8));
+
+	if (rc != 0) {
+		LOG_ERR("failed to write output port: %d", rc);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int pcal6416a_triggers_apply(const struct i2c_dt_spec *i2c,
+				    const struct pcal64xxa_triggers *triggers)
+{
+	int rc;
+	uint16_t input_latch = ~triggers->masked;
+	uint16_t interrupt_mask = triggers->masked;
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_INPUT_LATCH_0, (uint8_t)input_latch);
+	if (rc != 0) {
+		LOG_ERR("failed to configure input latch: %d", rc);
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_INPUT_LATCH_1, (uint8_t)(input_latch >> 8));
+	if (rc != 0) {
+		LOG_ERR("failed to configure input latch: %d", rc);
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_INTERRUPT_MASK_0, (uint8_t)interrupt_mask);
+	if (rc != 0) {
+		LOG_ERR("failed to configure interrupt mask: %d", rc);
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_INTERRUPT_MASK_1,
+				 (uint8_t)(interrupt_mask >> 8));
+	if (rc != 0) {
+		LOG_ERR("failed to configure interrupt mask: %d", rc);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int pcal6416a_reset_state_apply(const struct i2c_dt_spec *i2c)
```

Comment:
Please add the complementary `pcal6408a_reset_state_apply`

Reply:
done